[Diagnostics] Port name shadowing diagnostics

xedin opened the pull request:
Diagnose an attempt to reference a top-level name shadowed by
a local member e.g.

```swift
extension Sequence {
  func test() -> Int {
    return max(1, 2)
  }
}
```

Here `max` refers to a global function `max<T>(_: T, _: T)` in `Swift`
module and can only be accessed by adding `Swift.` to it, because `Sequence`
has a member named `max` which accepts a single argument.

Resolves: rdar://problem/25197358
Resolves: rdar://problem/30528057
Resolves: rdar://problem/32854314
Resolves: rdar://problem/58939370
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSFix] Add a fix to add a missing qualifier to shadowed top-level name references

xedin committed:
[Diagnostics] Port name shadowing diagnostics

Diagnose an attempt to reference a top-level name shadowed by
a local member e.g.

```swift
extension Sequence {
  func test() -> Int {
    return max(1, 2)
  }
}
```

Here `min` refers to a global function `min<T>(_: T, _: T)` in `Swift`
module and can only be accessed by adding `Swift.` to it, because `Sequence`
has a member named `min` which accepts a single argument.

xedin committed:
[ConstraintSystem] Use new fix/diagnostic for name shadowing

Stop filtering outer overload choices while trying to pre-check
expression, instead have it always fetch those and use new
fix to only attempt them in diagnostic mode (unless it's min/max
situation with conditional conformances).

xedin committed:
[CSDiag] NFC: Remove obsolete name shadowing diagnostics

xedin commented on lib/AST/NameLookup.cpp:
```diff
@@ -1450,6 +1450,16 @@ static bool isAcceptableLookupResult(const DeclContext *dc,
   }
 
+  // We don't support nested types in generics yet.
+  if (auto *NTD = dyn_cast<NominalTypeDecl>(decl)) {
```
@DougGregor @CodaFi I need some help with this since I don't really know what is the best place to filter these out, it seems like such members shouldn't be returned from `getMembers()` so they shouldn't loaded into a lookup table, but I couldn't figure out how member loading works exactly because none of the request break points I set were triggered.

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
I think this should be okay to use performance wise until we figure out a better solution by refactoring `resolveDeclRefExpr`. All of the "outer" results are going to be disabled by default.

theblixguy commented on lib/Sema/CSDiagnostics.h:
````diff
@@ -1924,6 +1924,29 @@ class UnableToInferClosureReturnType final : public FailureDiagnostic {
+/// }
+/// ```
+///
+/// Here `min` refers to a global function `min<T>(_: T, _: T)` in `Swift`
````
Typo: `min` should be `max` according to code snippet above

xedin replied:
Ugh, thank you!

xedin replied:
Fixed!

xedin committed:
[Diagnostics] NFC: Fix name shadowing diagnostic comment to refer to `max`

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please build toolchain macOS platform

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -1450,6 +1450,16 @@ static bool isAcceptableLookupResult(const DeclContext *dc,
     }
   }
 
+  // We don't support nested types in generics yet.
```
I don't think "in generics" is accurate; you're filtering out generic type members from protocols and local context instead. However it would be better to allow such results, since we want to lift the restriction eventually (in particular I want to allow generic types in local context soon), and instead fix whatever other crash you were seeing.

xedin replied:
I don't know whether it's possible to fix without implementing that feature or it would actually cause even more fallout... 

Here is a reduced example which cases a crash:

```
class A : P {
  let d = A
}

protocol P {
  class A {}
}
```

In this example `lookupUnqualified` returns one inner result for `P.A` and one outer one for `A`. The problem is that https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckNameLookup.cpp#L184L198 only handles associated type and protocol requirement cases. I was thinking about adding a `else` there to `{ addResult(found); return; }` but I'm not sure whether that's appropriate. 

This is yet another example of a situation which gets diagnosed by declaration checker and pushed on clients to handle (without even marking it as invalid).

Since I don't really know this area very well I'd be great to get an idea how do problems like this get fixed.

xedin replied:
Btw, I got "in generic" from `checkUnsupportedNestedType` check in `DeclChecker`...

slavapestov replied:
Ah, we want to get rid of that code eventually. Adding an 'else' there might work. If not, it looks like such results are simply dropped from the type checker lookup; which might be OK at that level too. I'm simply opposed to dropping them in the AST-level name lookup, since we shouldn't be enforcing semantic invariants there.

As for checkUnsupportedNestedType(), sounds like the comment there is out of date too :)

xedin replied:
Is adding that to LookupResultBuilder is fine, I'll gladly do that since it worked when I've testing it! :)

xedin commented:
@swift-ci please build toolchain macOS platform

xedin committed:
[TypeChecker] Add types nested in protocol to lookup results

Although such functionality is not yet supported we have to
mirror AST lookup and add such members into results, otherwise
there is a risk of inner/outer results mismatch.

xedin commented on test/decl/nested/protocol.swift:
```diff
@@ -32,7 +32,7 @@ struct ConformsToOuterProtocol : OuterProtocol {
 
   func f() { let _ = InnerProtocol.self }
-  // expected-error@-1 {{use of unresolved identifier 'InnerProtocol'}}
+  // expected-error@-1 {{protocol 'InnerProtocol' can only be used as a generic constraint because it has Self or associated type requirements}}
```
@slavapestov This is the only test change caused by adding nested types to result set in `LookupResultBuilder`.

slavapestov replied:
Excellent, thanks for following up!

xedin replied:
That’s for the help!

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test Linux platform

xedin commented:
There are a couple of LLDB failures:

```
12:53:33     lldb-Suite :: lang/swift/private_var/TestSwiftPrivateVar.py
12:53:33     lldb-Suite :: lang/swift/resilience/TestResilience.py
```

which have to do with unknown symbol for `unsafeMutableAddressor`. Don't seem related to my changes.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@dcci Could you please help me understand what effect `IncludeOuterResults` lookup flag has on LLDB?

The only test-case failing is `test/lang/swift/private_var/TestSwiftPrivateVar.py`:

```swift
private var a = 1
func doSomething(b: Int) {
	a += b //% self.expect("expr a", substrs=['Int', '= 1'])
}
doSomething(b:2)
```

And it fails this way:

```
20:41:35 runCmd: expr a
20:41:35 runCmd failed!
20:41:35 error: Couldn't lookup symbols:
20:41:35   a.(a in _8EC5D11DF8416F30E6CB4119BEB79E77).unsafeMutableAddressor : Swift.Int
20:41:35 
```

There seems to be no "shadowing" in that example - `a` is defined only once as a top-level variable so I have no idea why `IncludeOuterResult` might be affecting this at all...

adrian-prantl commented:
It's not immediately obvious what is going there, but since the expression evaluator has to inject itself into the symbol lookup process, I'm not surprised that there is an interaction. We'll have to look at this together in the debugger to fully understand what's going on. In the mean time you can reproduce the bug while having "lldb log enable expr" enabled, then you can see all the lookups being performed. Perhaps that gives you a clue.

adrian-prantl commented:
Here are the relevant pieces from the expression log:

```
 Parsing the following code:
@LLDBDebuggerFunction @available(OSX 10.15, *)
func $__lldb_expr(_ $__lldb_arg : UnsafeMutablePointer<Any>) {
do
{
/*__LLDB_USER_START__*/
a
/*__LLDB_USER_END__*/
}
catch (let __lldb_tmp_error)
{
    var $__lldb_error_result = __lldb_tmp_error
}
}
```

This is what is being passed to the type checker.

We're getting the following callbacks:
```
 [LLDBExprNameLookup::lookupOverrides(0)] Searching for UnsafeMutablePointer
 [LLDBExprNameLookup::lookupAdditions (0)] Searching for UnsafeMutablePointer
 [LLDBExprNameLookup::lookupOverrides(1)] Searching for ==
 [LLDBExprNameLookup::lookupAdditions (1)] Searching for ==
 [LLDBExprNameLookup::lookupOverrides(2)] Searching for IntegerLiteralType
 [LLDBExprNameLookup::lookupAdditions (2)] Searching for IntegerLiteralType
 [LLDBExprNameLookup::lookupOverrides(3)] Searching for FloatLiteralType
 [LLDBExprNameLookup::lookupAdditions (3)] Searching for FloatLiteralType

Source file after type checking:
 (source_file "<EXPR>"
  (func_decl range=[<EXPR>:2:1 - line:13:1] "$__lldb_expr(_:)" interface type='(UnsafeMutablePointer<Any>) -> ()' access=internal
    (parameter_list
      (parameter "$__lldb_arg" type='UnsafeMutablePointer<Any>' interface type='UnsafeMutablePointer<Any>') range=[<EXPR>:2:18 - line:2:60])
    (brace_stmt range=[<EXPR>:2:62 - line:13:1]
      (var_decl implicit range=[<EXPR>:2:62 - line:2:62] "a" type='Int' interface type='Int' access=private readImpl=stored writeImpl=stored readWriteImpl=stored)

      (pattern_binding_decl implicit range=[<EXPR>:2:62 - line:2:62]
        (pattern_typed implicit type='Int'
          (pattern_named implicit type='Int' 'a')))
...

Variables:
   [name=b, type = Swift.Int]
   [name=a, type = Swift.Int]
   [name=$__lldb_result, type = Swift.Int, is_result]
   [name=$__lldb_error_result, type = Swift.Error, is_error]
```

Does this make things more clear, or more confusing?

xedin commented:
@adrian-prantl Thank you! I think that makes it more confusing as to where `unsafeMutableAddressor` comes into play at all...

xedin commented:
@swift-ci please test compiler performance

adrian-prantl committed:
Serialize whether a VarDecl is a top-level global.

The current way that VarDecl::isLazilyInitializedGlobal() is implemented does
not work in the debugger, since the DeclContext of all VarDecls are deserialized
Swift modules. By adding a bit to the VarDecl we can recover the fact that a
VarDecl was in fact a global even in the debugger.

<rdar://problem/58939370>

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test

xedin commented:
@adrian-prantl Looks like some of the tests produce different output now...

adrian-prantl committed:
Compile libraries in testcases with -parse-as-library (NFC)

This is in preparation to a change in serialization of global variables where
this detail will matter.

xedin commented:
test with apple/llvm-project#674
@swift-ci test

xedin commented:
test with apple/llvm-project#674
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test compiler performance

xedin commented:
# Summary for master full

**Unexpected test results, excluded stats for RxCocoa, SwifterSwift, Base64CoderSwiftUI**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-4157781eb7db981f)
    - [debug-batch brief](#debug-batch-brief-4157781eb7db981f)
    - [debug-batch detailed](#debug-batch-detailed-4157781eb7db981f)
- [Release](#release-4157781eb7db981f)
    - [release brief](#release-brief-4157781eb7db981f)
    - [release detailed](#release-detailed-4157781eb7db981f)



<a name="debug-batch-4157781eb7db981f">

# Debug-batch

<a name="debug-batch-brief-4157781eb7db981f">

## debug-batch brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 79,946,409,430,760 | 79,817,329,963,083 | -129,079,467,677 | -0.16%
LLVM.NumLLVMBytesOutput | 1,822,118,222 | 1,822,117,268 | -954 | -0.0%
time.swift-driver.wall | 6241.6s | 6234.8s | -6.7s | -0.11%
</details>

<a name="debug-batch-detailed-4157781eb7db981f">

## debug-batch detailed
<details>
<summary>Regressed (8)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.ImportSetCacheHit | 1,945,753 | 2,169,858 | 224,105 | 11.52% :no_entry:
AST.ModuleShadowCacheHit | 2,356 | 3,602 | 1,246 | 52.89% :no_entry:
AST.ModuleShadowCacheMiss | 1,667 | 1,913 | 246 | 14.76% :no_entry:
AST.NumModuleLookupValue | 49,775,293 | 54,983,428 | 5,208,135 | 10.46% :no_entry:
AST.NumReferencedTopLevelNames | 484,793 | 565,430 | 80,637 | 16.63% :no_entry:
Sema.DefaultTypeRequest | 480,641 | 485,948 | 5,307 | 1.1% :no_entry:
Sema.LookupInModuleRequest | 7,068,252 | 7,293,583 | 225,331 | 3.19% :no_entry:
Sema.TypeWitnessRequest | 9,657 | 9,906 | 249 | 2.58% :no_entry:
</details>
<details>
<summary>Improved (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Driver.NumDriverPipePolls | 80,516 | 79,156 | -1,360 | -1.69% :white_check_mark:
Driver.NumDriverPipeReads | 67,897 | 66,611 | -1,286 | -1.89% :white_check_mark:
Sema.ExistentialConformsToSelfRequest | 17,258 | 14,478 | -2,780 | -16.11% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (199)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.ImportSetCacheMiss | 1,044,754 | 1,044,718 | -36 | -0.0%
AST.ImportSetFoldHit | 563,437 | 563,430 | -7 | -0.0%
AST.ImportSetFoldMiss | 481,316 | 481,287 | -29 | -0.01%
AST.ModuleVisibilityCacheHit | 34,969 | 35,063 | 94 | 0.27%
AST.ModuleVisibilityCacheMiss | 9,151 | 9,157 | 6 | 0.07%
AST.NumASTBytesAllocated | 72,079,702,413 | 72,005,122,198 | -74,580,215 | -0.1%
AST.NumASTScopeLookups | 4,397,979 | 4,401,631 | 3,652 | 0.08%
AST.NumBraceStmtASTScopeExpansions | 639,722 | 639,722 | 0 | 0.0%
AST.NumBraceStmtASTScopes | 639,722 | 639,722 | 0 | 0.0%
AST.NumDecls | 143,587 | 143,587 | 0 | 0.0%
AST.NumDependencies | 448,517 | 448,585 | 68 | 0.02%
AST.NumInfixOperators | 54,567 | 54,567 | 0 | 0.0%
AST.NumIterableTypeBodyASTScopeExpansions | 278,303 | 278,370 | 67 | 0.02%
AST.NumIterableTypeBodyASTScopes | 365,167 | 365,149 | -18 | -0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 414,192 | 414,192 | 0 | 0.0%
AST.NumLocalTypeDecls | 254 | 254 | 0 | 0.0%
AST.NumModuleLookupClassMember | 6,927 | 6,927 | 0 | 0.0%
AST.NumObjCMethods | 24,747 | 24,747 | 0 | 0.0%
AST.NumPostfixOperators | 49 | 49 | 0 | 0.0%
AST.NumPrecedenceGroups | 26,329 | 26,329 | 0 | 0.0%
AST.NumPrefixOperators | 99 | 99 | 0 | 0.0%
AST.NumReferencedDynamicNames | 189 | 189 | 0 | 0.0%
AST.NumReferencedMemberNames | 6,437,299 | 6,469,647 | 32,348 | 0.5%
AST.NumSourceBuffers | 1,033,777 | 1,033,788 | 11 | 0.0%
AST.NumSourceLines | 4,795,108 | 4,795,108 | 0 | 0.0%
AST.NumSourceLinesPerSecond | 3,264,694 | 3,272,358 | 7,664 | 0.23%
AST.NumTotalClangImportedEntities | 6,361,738 | 6,361,901 | 163 | 0.0%
Driver.ChildrenMaxRSS | 252,445,878,272 | 251,975,567,360 | -470,310,912 | -0.19%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 28,508 | 28,508 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 1,493,756,096,416 | 1,492,964,903,288 | -791,193,128 | -0.05%
Frontend.NumInstructionsExecuted | 79,946,409,430,760 | 79,817,329,963,083 | -129,079,467,677 | -0.16%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 199,469 | 199,469 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 6,985,707 | 6,985,707 | 0 | 0.0%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 3,437,768 | 3,437,768 | 0 | 0.0%
IRModule.NumIRGlobals | 3,735,700 | 3,735,700 | 0 | 0.0%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 88,378,438 | 88,378,438 | 0 | 0.0%
IRModule.NumIRNamedMetaData | 137,660 | 137,660 | 0 | 0.0%
IRModule.NumIRValueSymbols | 6,508,017 | 6,508,017 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,822,118,222 | 1,822,117,268 | -954 | -0.0%
Parse.NumFunctionsParsed | 273,969 | 273,969 | 0 | 0.0%
Parse.NumIterableDeclContextParsed | 962,809 | 963,579 | 770 | 0.08%
Parse.ParseAbstractFunctionBodyRequest | 250,031 | 250,031 | 0 | 0.0%
Parse.ParseMembersRequest | 791,614 | 792,384 | 770 | 0.1%
SILGen.GenerateSILRequest | 28,497 | 28,497 | 0 | 0.0%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 1,733,138 | 1,733,138 | 0 | 0.0%
SILModule.NumSILGenGlobalVariables | 54,724 | 54,724 | 0 | 0.0%
SILModule.NumSILGenVtables | 18,938 | 18,938 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 73,908 | 73,908 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 2,558,124 | 2,558,124 | 0 | 0.0%
SILModule.NumSILOptGlobalVariables | 56,390 | 56,390 | 0 | 0.0%
SILModule.NumSILOptVtables | 32,041 | 32,041 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 171,962 | 171,962 | 0 | 0.0%
Sema.AbstractGenericSignatureRequest | 35,189 | 35,189 | 0 | 0.0%
Sema.AccessLevelRequest | 14,694,952 | 14,658,099 | -36,853 | -0.25%
Sema.AnyObjectLookupRequest | 281 | 281 | 0 | 0.0%
Sema.AreAllStoredPropertiesDefaultInitableRequest | 23,540 | 23,540 | 0 | 0.0%
Sema.AttachedFunctionBuilderRequest | 1 | 1 | 0 | 0.0%
Sema.AttachedPropertyWrapperTypeRequest | 550,960 | 550,960 | 0 | 0.0%
Sema.AttachedPropertyWrappersRequest | 2,423,743 | 2,423,734 | -9 | -0.0%
Sema.CallerSideDefaultArgExprRequest | 78,971 | 78,971 | 0 | 0.0%
Sema.ClassAncestryFlagsRequest | 113,887 | 113,887 | 0 | 0.0%
Sema.CollectOverriddenDeclsRequest | 9,078,822 | 9,044,315 | -34,507 | -0.38%
Sema.CompareDeclSpecializationRequest | 488,947 | 488,494 | -453 | -0.09%
Sema.CursorInfoRequest | 0 | 0 | 0 | 0.0%
Sema.CustomAttrNominalRequest | 1 | 1 | 0 | 0.0%
Sema.DefaultAndMaxAccessLevelRequest | 59,752 | 59,764 | 12 | 0.02%
Sema.DefaultArgumentExprRequest | 39,319 | 39,319 | 0 | 0.0%
Sema.DefaultArgumentInitContextRequest | 458 | 458 | 0 | 0.0%
Sema.DefaultDefinitionTypeRequest | 8,411 | 8,411 | 0 | 0.0%
Sema.DifferentiableAttributeTypeCheckRequest | 0 | 0 | 0 | 0.0%
Sema.DirectLookupRequest | 35,375,554 | 35,467,709 | 92,155 | 0.26%
Sema.DynamicallyReplacedDeclRequest | 1,089,887 | 1,089,887 | 0 | 0.0%
Sema.EmittedMembersRequest | 28,512 | 28,512 | 0 | 0.0%
Sema.EnumRawTypeRequest | 26,378 | 26,380 | 2 | 0.01%
Sema.EnumRawValuesRequest | 10,748 | 10,748 | 0 | 0.0%
Sema.ExistentialTypeSupportedRequest | 18,355 | 18,355 | 0 | 0.0%
Sema.ExpandASTScopeRequest | 7,847,080 | 7,850,498 | 3,418 | 0.04%
Sema.ExtendedNominalRequest | 942,165 | 942,165 | 0 | 0.0%
Sema.ExtendedTypeRequest | 86,281 | 86,340 | 59 | 0.07%
Sema.FunctionBuilderTypeRequest | 1 | 1 | 0 | 0.0%
Sema.FunctionOperatorRequest | 74,104 | 74,104 | 0 | 0.0%
Sema.GenericParamListRequest | 11,863,808 | 11,842,553 | -21,255 | -0.18%
Sema.GenericSignatureRequest | 3,355,599 | 3,357,290 | 1,691 | 0.05%
Sema.GetDestructorRequest | 28,919 | 28,919 | 0 | 0.0%
Sema.HasCircularInheritanceRequest | 24,294 | 24,294 | 0 | 0.0%
Sema.HasCircularInheritedProtocolsRequest | 10,466 | 10,466 | 0 | 0.0%
Sema.HasCircularRawValueRequest | 9,127 | 9,127 | 0 | 0.0%
Sema.HasDefaultInitRequest | 58,039 | 58,039 | 0 | 0.0%
Sema.HasDynamicMemberLookupAttributeRequest | 677,187 | 677,140 | -47 | -0.01%
Sema.HasMemberwiseInitRequest | 20,721 | 20,721 | 0 | 0.0%
Sema.HasMissingDesignatedInitializersRequest | 27,324 | 27,324 | 0 | 0.0%
Sema.HasUserDefinedDesignatedInitRequest | 58,050 | 58,050 | 0 | 0.0%
Sema.InferredGenericSignatureRequest | 193,792 | 194,174 | 382 | 0.2%
Sema.InheritedDeclsReferencedRequest | 7,480,128 | 7,473,011 | -7,117 | -0.1%
Sema.InheritedTypeRequest | 323,150 | 324,782 | 1,632 | 0.51%
Sema.InheritsSuperclassInitializersRequest | 30,498 | 30,498 | 0 | 0.0%
Sema.InitKindRequest | 100,173 | 100,173 | 0 | 0.0%
Sema.InterfaceTypeRequest | 20,519,461 | 20,485,182 | -34,279 | -0.17%
Sema.IsABICompatibleOverrideRequest | 138,436 | 138,436 | 0 | 0.0%
Sema.IsAccessorTransparentRequest | 324,533 | 324,533 | 0 | 0.0%
Sema.IsCallableNominalTypeRequest | 2,132 | 2,132 | 0 | 0.0%
Sema.IsDeclApplicableRequest | 0 | 0 | 0 | 0.0%
Sema.IsDynamicRequest | 1,786,375 | 1,786,419 | 44 | 0.0%
Sema.IsFinalRequest | 2,885,109 | 2,881,858 | -3,251 | -0.11%
Sema.IsGetterMutatingRequest | 443,646 | 443,653 | 7 | 0.0%
Sema.IsImplicitlyUnwrappedOptionalRequest | 2,580,321 | 2,581,105 | 784 | 0.03%
Sema.IsObjCRequest | 1,700,824 | 1,700,542 | -282 | -0.02%
Sema.IsSetterMutatingRequest | 349,588 | 349,588 | 0 | 0.0%
Sema.IsStaticRequest | 3,759,140 | 3,758,049 | -1,091 | -0.03%
Sema.LazyStoragePropertyRequest | 2,570 | 2,570 | 0 | 0.0%
Sema.LookupPrecedenceGroupRequest | 179,608 | 179,608 | 0 | 0.0%
Sema.MangleLocalTypeDeclRequest | 508 | 508 | 0 | 0.0%
Sema.ModuleQualifiedLookupRequest | 3,162,425 | 3,160,133 | -2,292 | -0.07%
Sema.NamedLazyMemberLoadFailureCount | 15,471 | 15,467 | -4 | -0.03%
Sema.NamedLazyMemberLoadSuccessCount | 12,976,338 | 13,005,645 | 29,307 | 0.23%
Sema.NamingPatternRequest | 212,939 | 212,962 | 23 | 0.01%
Sema.NeedsNewVTableEntryRequest | 708,793 | 708,793 | 0 | 0.0%
Sema.NumAccessorBodiesSynthesized | 191,878 | 191,878 | 0 | 0.0%
Sema.NumAccessorsSynthesized | 300,881 | 300,881 | 0 | 0.0%
Sema.NumConformancesDeserialized | 10,746,327 | 10,731,261 | -15,066 | -0.14%
Sema.NumConstraintScopes | 28,893,540 | 28,913,975 | 20,435 | 0.07%
Sema.NumConstraintsConsideredForEdgeContraction | 98,520,334 | 98,540,647 | 20,313 | 0.02%
Sema.NumCyclicOneWayComponentsCollapsed | 0 | 0 | 0 | 0.0%
Sema.NumDeclsDeserialized | 85,022,865 | 84,923,583 | -99,282 | -0.12%
Sema.NumDeclsTypechecked | 1,437,871 | 1,437,915 | 44 | 0.0%
Sema.NumFunctionsTypechecked | 538,730 | 538,730 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 1,637,474 | 1,637,244 | -230 | -0.01%
Sema.NumLazyIterableDeclContexts | 10,351,098 | 10,347,176 | -3,922 | -0.04%
Sema.NumLazyRequirementSignatures | 1,080,655 | 1,080,595 | -60 | -0.01%
Sema.NumLazyRequirementSignaturesLoaded | 799,266 | 799,229 | -37 | -0.0%
Sema.NumLeafScopes | 18,591,716 | 18,608,311 | 16,595 | 0.09%
Sema.NumTypesDeserialized | 25,398,885 | 25,386,733 | -12,152 | -0.05%
Sema.NumTypesValidated | 1,554,366 | 1,556,662 | 2,296 | 0.15%
Sema.NumUnloadedLazyIterableDeclContexts | 6,401,325 | 6,403,169 | 1,844 | 0.03%
Sema.OpaqueReadOwnershipRequest | 287,508 | 287,502 | -6 | -0.0%
Sema.OpaqueResultTypeRequest | 0 | 0 | 0 | 0.0%
Sema.OperatorPrecedenceGroupRequest | 747 | 747 | 0 | 0.0%
Sema.OverriddenDeclsRequest | 2,723,618 | 2,720,669 | -2,949 | -0.11%
Sema.ParamSpecifierRequest | 1,331,237 | 1,333,467 | 2,230 | 0.17%
Sema.PatternBindingEntryRequest | 474,569 | 474,589 | 20 | 0.0%
Sema.PatternTypeRequest | 598,446 | 598,466 | 20 | 0.0%
Sema.PreCheckFunctionBuilderRequest | 0 | 0 | 0 | 0.0%
Sema.PropertyWrapperBackingPropertyInfoRequest | 545,513 | 545,513 | 0 | 0.0%
Sema.PropertyWrapperBackingPropertyTypeRequest | 550,960 | 550,960 | 0 | 0.0%
Sema.PropertyWrapperMutabilityRequest | 643,545 | 643,552 | 7 | 0.0%
Sema.PropertyWrapperTypeInfoRequest | 7 | 7 | 0 | 0.0%
Sema.ProtocolRequiresClassRequest | 77,452 | 77,421 | -31 | -0.04%
Sema.ProvideDefaultImplForRequest | 9,078,822 | 9,044,315 | -34,507 | -0.38%
Sema.QualifiedLookupRequest | 7,542,968 | 7,565,080 | 22,112 | 0.29%
Sema.RangeInfoRequest | 0 | 0 | 0 | 0.0%
Sema.RequirementRequest | 119,395 | 120,117 | 722 | 0.6%
Sema.RequirementSignatureRequest | 880,770 | 880,757 | -13 | -0.0%
Sema.RequiresOpaqueAccessorsRequest | 1,341,327 | 1,341,327 | 0 | 0.0%
Sema.RequiresOpaqueModifyCoroutineRequest | 278,072 | 278,072 | 0 | 0.0%
Sema.ResilienceExpansionRequest | 1,903,354 | 1,904,113 | 759 | 0.04%
Sema.ResolveImplicitMemberRequest | 407,323 | 407,181 | -142 | -0.03%
Sema.ResolveProtocolNameRequest | 0 | 0 | 0 | 0.0%
Sema.ResultTypeRequest | 708,087 | 708,714 | 627 | 0.09%
Sema.RootAndResultTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.RootTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.SelfAccessKindRequest | 937,149 | 936,390 | -759 | -0.08%
Sema.SelfBoundsFromWhereClauseRequest | 1,906,389 | 1,905,942 | -447 | -0.02%
Sema.SetterAccessLevelRequest | 147,874 | 147,874 | 0 | 0.0%
Sema.StorageImplInfoRequest | 1,453,713 | 1,453,715 | 2 | 0.0%
Sema.StoredPropertiesAndMissingMembersRequest | 33,206 | 33,206 | 0 | 0.0%
Sema.StoredPropertiesRequest | 378,155 | 378,155 | 0 | 0.0%
Sema.StructuralTypeRequest | 2,336 | 2,336 | 0 | 0.0%
Sema.SuperclassDeclRequest | 520,189 | 520,110 | -79 | -0.02%
Sema.SuperclassTypeRequest | 61,701 | 61,701 | 0 | 0.0%
Sema.SynthesizeAccessorRequest | 300,881 | 300,881 | 0 | 0.0%
Sema.SynthesizeDefaultInitRequest | 5,421 | 5,421 | 0 | 0.0%
Sema.SynthesizeMemberwiseInitRequest | 3,247 | 3,247 | 0 | 0.0%
Sema.TypeCheckFunctionBodyUntilRequest | 538,730 | 538,730 | 0 | 0.0%
Sema.TypeCheckSourceFileRequest | 27,532 | 27,532 | 0 | 0.0%
Sema.TypeDeclsFromWhereClauseRequest | 31,205 | 31,217 | 12 | 0.04%
Sema.TypeRelationCheckRequest | 0 | 0 | 0 | 0.0%
Sema.USRGenerationRequest | 29,341,294 | 29,248,093 | -93,201 | -0.32%
Sema.UnderlyingTypeDeclsReferencedRequest | 406,389 | 406,244 | -145 | -0.04%
Sema.UnderlyingTypeRequest | 39,284 | 39,330 | 46 | 0.12%
Sema.UnqualifiedLookupRequest | 4,545,114 | 4,548,766 | 3,652 | 0.08%
Sema.ValueWitnessRequest | 70,304 | 70,316 | 12 | 0.02%
</details>

<a name="release-4157781eb7db981f">

# Release

<a name="release-brief-4157781eb7db981f">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 49,394,321,687,141 | 49,358,978,208,522 | -35,343,478,619 | -0.07%
LLVM.NumLLVMBytesOutput | 1,862,180,780 | 1,862,198,624 | 17,844 | 0.0%
time.swift-driver.wall | 9063.8s | 9050.9s | -13.0s | -0.14%
</details>

<a name="release-detailed-4157781eb7db981f">

## release detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (18)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumLoadedModules | 30,691 | 30,691 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 1,152,447 | 1,155,245 | 2,798 | 0.24%
IRModule.NumIRBasicBlocks | 5,035,260 | 5,035,260 | 0 | 0.0%
IRModule.NumIRFunctions | 2,828,694 | 2,828,694 | 0 | 0.0%
IRModule.NumIRGlobals | 3,256,565 | 3,256,565 | 0 | 0.0%
IRModule.NumIRInsts | 50,642,167 | 50,642,167 | 0 | 0.0%
IRModule.NumIRValueSymbols | 5,720,660 | 5,720,660 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,862,180,780 | 1,862,198,624 | 17,844 | 0.0%
SILModule.NumSILGenFunctions | 1,212,481 | 1,212,481 | 0 | 0.0%
SILModule.NumSILOptFunctions | 1,027,129 | 1,027,129 | 0 | 0.0%
Sema.NumConformancesDeserialized | 3,528,797 | 3,528,955 | 158 | 0.0%
Sema.NumConstraintScopes | 28,428,076 | 28,450,539 | 22,463 | 0.08%
Sema.NumDeclsDeserialized | 10,868,715 | 10,872,555 | 3,840 | 0.04%
Sema.NumFunctionsTypechecked | 542,352 | 542,352 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 285,897 | 285,906 | 9 | 0.0%
Sema.NumLazyIterableDeclContexts | 1,345,916 | 1,345,984 | 68 | 0.01%
Sema.NumTypesDeserialized | 5,267,948 | 5,270,754 | 2,806 | 0.05%
Sema.NumTypesValidated | 841,795 | 841,795 | 0 | 0.0%
</details>

xedin commented:
Looks like the only counters regressed are associated with debug configuration and expected as they are all about global requests which with `IncludeOuterResult` we definitely do more of. But wall time didn't regress, so it looks like we can go ahead and merge these changes after I made a small change @slavapestov has requested.

xedin commented:
Ok, `VarDecl::isLazilyInitializedGlobal()` simplification is in and I'm going to run CI one more time.

xedin commented:
@swift-ci please test

xedin commented:
test with apple/llvm-project#674
@swift-ci please test

xedin commented:
test with apple/llvm-project#674
@swift-ci please test

xedin commented:
test with apple/llvm-project#674
@swift-ci please test
